server/src/main/java/org/elasticsearch/monitor/fs/FsHealthService.java

```diff
@@ -0,0 +1,175 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.monitor.fs;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.common.component.AbstractLifecycleComponent;
+import org.elasticsearch.common.settings.ClusterSettings;
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.core.internal.io.IOUtils;
+import org.elasticsearch.env.NodeEnvironment;
+import org.elasticsearch.monitor.NodeHealthService;
+import org.elasticsearch.threadpool.Scheduler;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardOpenOption;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Random;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.LongSupplier;
+
+/**
+ * Runs periodically and attempts to create a temp file to see if the filesystem is writable. If not then it marks the
+ * path as unhealthy.
+ */
+public class FsHealthService extends AbstractLifecycleComponent implements NodeHealthService {
+
+    private static final Logger logger = LogManager.getLogger(FsHealthService.class);
```

Comment:
I would like us to log when we become unhealthy in this service rather than relying on its clients logging warnings/errors, since I am concerned that the clients' messages may be hard to interpret and they vary a lot which will make analysis a good deal harder. That means I think we'll need to schedule a timeout handler for each check. It would be mightily useful if that log message were to include which path(s) were faulty, and how long it took to check each path.

I'd also like us to log when a check takes longer than some threshold (say, `10s` by default) but not long enough to fail the node. This kind of sub-threshold warning can be a useful leading indicator of impending doom. See e.g. `cluster.publish.info_timeout` vs `cluster.publish.timeout`.

Reply:
Should we consider multiple logging levels here like separate thresholds for info/warn levels?

Reply:
I don't think we need multiple warnings before hitting the threshold for failure, but I think it's reasonable to log the warning at `WARN` and then log the actual failure at `ERROR`.